lib/generate-critical-css.js

```diff
@@ -1,26 +1,84 @@
 const CSSFileSet = require( './css-file-set' );
 const { removeIgnoredPseudoElements } = require( './ignored-pseudo-elements' );
 const { minifyCss } = require( './minify-css' );
+const { SuccessTargetError } = require( './errors' );
+const { BrowserInterface } = require( './browser-interface' );
+
+/**
+ * Collate and return a CSSFileSet object describing all the CSS files used by
+ * the set of URLs provided.
+ *
+ * URLs which fail to load generate warnings, but do not stop the process.
+ *
+ * @param {BrowserInterface} browserInterface - interface to access pages
+ * @param {string[]} urls - list of URLs to scan for CSS files
+ * @param {number} successUrlsThreshold - success urls amount threshold
+ * @return {CSSFileSet} - Set of CSS files used by the urls.
+ */
+async function collateCssFiles( browserInterface, urls, successUrlsThreshold ) {
+	const cssFiles = new CSSFileSet();
+	const warnings = [];
+	let successes = 0;
+
+	for ( const url of urls ) {
+		try {
+			const cssIncludes = await browserInterface.getCssIncludes( url );
+
+			// Convert relative URLs to absolute.
+			const relativeUrls = Object.keys( cssIncludes );
+			const absoluteIncludes = relativeUrls.reduce( ( set, relative ) => {
+				const absolute = new URL( relative, url ).toString();
+				set[ absolute ] = cssIncludes[ relative ];
+
+				return set;
+			}, {} );
+
+			await cssFiles.addMultiple( url, absoluteIncludes );
+
+			// Abort early if we hit the threshold of success urls.
+			successes++;
+			if ( successes >= successUrlsThreshold ) {
+				break;
+			}
+		} catch ( err ) {
+			warnings.push( err );
+		}
+	}
+
+	return [ cssFiles, warnings ];
+}
 
 async function generateCriticalCSS( {
 	browserInterface,
 	progressCallback,
 	urls,
 	viewports,
 	filters,
+	successRatio = 1,
 } ) {
-	if ( ! progressCallback ) {
-		progressCallback = () => {};
-	}
+	// TODO: Review if we want to be less strict and use Math.floor instead.
```

Comment:
@thingalon , @pyronaur : I am not sure if we want to be more conservative using `ceil` which means that the success urls threshold is one url more, or a little more tolerant and use `floor`

Reply:
We certainly don't want to use `Math.floor` here. If there is 1 URL in the set, then it will always set the threshold at 0 ;)

Reply:
Sounds good. Let's stick with `ceil` :)